src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+                final Object oldValue = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+                Object newValue = callBlock(block, oldValue);
+
+                if (newValue == nil) {
+                    return null;
+                } else {
+                    return newValue;
+                }
+            });
+
+            if (returnValue == null) {
+                return nil;
+            }
+            return returnValue;
+        }
+    }
+
+    @CoreMethod(names = "replace_pair", required = 3)
+    public abstract static class ReplacePairNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected boolean replacePair(RubyConcurrentHashMap self, Object key, Object oldValue, Object newValue) {
+            return self.concurrentHash.replace(new RubyConcurrentHashMap.Key(key), oldValue, newValue);
+        }
+    }
+
+    @CoreMethod(names = "replace_if_exists", required = 2)
+    public abstract static class ReplaceIfExistsNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object replaceIfExists(RubyConcurrentHashMap self, Object key, Object newValue) {
+            final Object oldValue = self.concurrentHash.replace(new RubyConcurrentHashMap.Key(key), newValue);
+            if (oldValue == null) {
+                return nil;
+            } else {
+                return oldValue;
+            }
+        }
+    }
+
+    @CoreMethod(names = "get_and_set", required = 2)
+    public abstract static class GetAndSetNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object getAndSet(RubyConcurrentHashMap self, Object key, Object value) {
+            final Object oldValue = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+            self.concurrentHash.put(new RubyConcurrentHashMap.Key(key), value);
+            if (oldValue == null) {
+                return nil;
+            } else {
+                return oldValue;
+            }
+        }
+    }
+
+    @CoreMethod(names = "key?", required = 1)
+    public abstract static class KeyNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected boolean key(RubyConcurrentHashMap self, Object key) {
+            return self.concurrentHash.containsKey(new RubyConcurrentHashMap.Key(key));
+        }
+    }
+
+    @CoreMethod(names = "delete", required = 1)
+    public abstract static class DeleteNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object delete(RubyConcurrentHashMap self, Object key) {
+            if (self.concurrentHash.containsKey(new RubyConcurrentHashMap.Key(key))) {
+                return self.concurrentHash.remove(new RubyConcurrentHashMap.Key(key));
```

Comment:
This is not atomic (two operations), use the return value of remove() instead (!= null means it was removed)

Reply:
`#delete` made atomic in [`0e5e8d3` (#2339)](https://github.com/oracle/truffleruby/pull/2339/commits/0e5e8d3fe44b100d5185d762e75414619d17aa51)